heartbeat/monitors/wrappers/monitorstate/monitorstate.go

```diff
@@ -0,0 +1,159 @@
+		// Note that we add an incrementing counter to help with the fact that
+		// millisecond res isn't quite enough for uniqueness (esp. in tests)
+		ID:         fmt.Sprintf("%s-%x-%x", monitorId, now.UnixMilli(), ctr),
+		StartedAt:  now,
+		DurationMs: 0,
+		Status:     status,
+		ctr:        ctr + 1,
+	}
+	ms.recordCheck(monitorId, status)
+
+	return ms
+}
+
+type State struct {
+	ID string `json:"id"`
+	// StartedAt is the start time of the state, should be the same for a given state ID
+	StartedAt  time.Time   `json:"started_at"`
+	DurationMs int64       `json:"duration_ms"`
+	Status     StateStatus `json:"status"`
+	Checks     int         `json:"checks"`
+	Up         int         `json:"up"`
+	Down       int         `json:"down"`
+	// FlapHistory retains enough info so we can resume our flap
+	// computation if loading from ES or another source
+	FlapHistory []StateStatus `json:"flap_history"`
+	// Ends is a pointer to the prior state if this is the start of a new state
+	Ends *State `json:"ends"`
+	ctr  int
+}
+
+func (s *State) incrementCounters(status StateStatus) {
+	s.DurationMs = time.Since(s.StartedAt).Milliseconds()
+	s.Checks++
+	if status == StatusUp {
+		s.Up++
+	} else {
+		s.Down++
+	}
+}
+
+// truncate flap history to be at most as many items as the threshold indicates, minus one
+func (s *State) truncateFlapHistory() {
+	endIdx := len(s.FlapHistory)
+	if endIdx < 0 {
+		return // flap history is empty
+	}
+	// truncate to one less than the threshold since our later calculations
+	// an item that would stabilize the history at the threshold would start a new state
+	startIdx := endIdx - (FlappingThreshold - 1)
+	if startIdx <= 0 {
+		return
+	}
+	s.FlapHistory = s.FlapHistory[startIdx:endIdx]
+}
+
+// recordCheck updates the current state pointer to what the new state should be.
+// If the current state is continued it just updates counters and other record keeping,
+// if the state ends it actually swaps out the full value the state points to
+// and sets state.Ends.
+func (s *State) recordCheck(monitorID string, newStatus StateStatus) {
+	if s.Status == StatusFlapping {
+		s.truncateFlapHistory()
+
+		// Check if all statuses in flap history are identical, including the new status
+		hasStabilized := true
+		for _, histStatus := range s.FlapHistory {
+			if newStatus != histStatus {
+				hasStabilized = false
+				break
+			}
+		}
+
+		if !hasStabilized { // continue flapping
+			// Use the new flap history as part of the state
+			s.FlapHistory = append(s.FlapHistory, newStatus)
+			s.incrementCounters(newStatus)
+		} else { // flap has ended
+			s.transitionTo(monitorID, newStatus)
+		}
+	} else if s.Status == newStatus { // stable state, status has not changed
+		// The state is stable, no changes needed
+		s.incrementCounters(newStatus)
+	} else if s.Checks < FlappingThreshold {
+		// The state changed too quickly, we're now flapping
+		s.incrementCounters(newStatus)
+		s.Status = StatusFlapping
```

Comment:
It seems like all monitors will have a period of time where they are considered flapping in this logic? Will all monitors, once the state changes from up to down or down to up, be considered flapping until the `FLAPPING_THRESHOLD` number of checks with the new status?

Reply:
It doesn't actually work that way (which can be seen in the real world data where we are more often than not in up or down, but not flap states), the condition above where `s.Status == newStatus` ensures that. In english I'd say that with `s.Checks < FlappingThreshold` we're in the eligible to flap zone.